src/Storages/MergeTree/MergeMutateExecutor.cpp

```diff
@@ -0,0 +1,147 @@
+#include <Storages/MergeTree/MergeMutateExecutor.h>
+
+#include <Storages/MergeTree/BackgroundJobsExecutor.h>
+
+
+namespace DB
+{
+
+
+/// This is a RAII class which only decrements metric.
+/// It is added because after all other fixes a bug non-executing merges was occurred again.
+/// Last hypothesis: task was successfully added to pool, however, was not executed because of internal exception in it.
+class ParanoidMetricDecrementor
+{
+public:
+    explicit ParanoidMetricDecrementor(CurrentMetrics::Metric metric_) : metric(metric_) {}
+    void alarm() { is_alarmed = true; }
+    void decrement()
+    {
+        if (is_alarmed.exchange(false))
+        {
+            CurrentMetrics::values[metric]--;
+        }
+    }
+
+    ~ParanoidMetricDecrementor() { decrement(); }
+
+private:
+
+    CurrentMetrics::Metric metric;
+    std::atomic_bool is_alarmed = false;
+};
+
+
+void MergeTreeBackgroundExecutor::removeTasksCorrespondingToStorage(StorageID id)
+{
+    std::lock_guard remove_lock(remove_mutex);
+
+    /// First stop the scheduler thread
+    {
+        std::unique_lock lock(mutex);
+        shutdown_suspend = true;
+        has_tasks.notify_one();
+    }
+
+    scheduler.join();
+
+    /// Remove tasks
+    {
+        std::lock_guard lock(currently_executing_mutex);
+
+        for (auto & [task, future] : currently_executing)
+        {
+            if (task->getStorageID() == id)
+                future.wait();
+        }
+
+        /// Remove tasks from original queue
+        size_t erased_count = std::erase_if(tasks, [id = std::move(id)] (auto task) -> bool { return task->getStorageID() == id; });
```

Comment:
That's suspicious.
Usually we access tasks under `mutex`. here we don't , and access from pool's to tasks is also possible.
btw can't we do it after set `shutdown_suspend = true;` in the same function?

Reply:
Yes, this is a bug (regarding mutex usage). But we can't first remove tasks from queue. Because threads from thread pool could execute some tasks related to current StorageID (which we want to delete) and some of this tasks could be placed into the queue after execution (some of them don't). So cleaning the queue at the end of function is intentionally.